### ISSUE TITLE
VScode testing tree doesn't list additional sections

### ISSUE THREAD

**Issue by cidrblock** [The-Compiler/vscode-python-tox#33]
I don't see this documented anywhere, so I'm not sure if it is intended to work.

Give a tox.ini in a directory, the testing tree will show specific entries in the ini but not everything environment available (e.f. `tox -l`)

![image](https://user-images.githubusercontent.com/18386516/231787622-52d9e5ec-08e3-4788-af62-4f990b759262.png)

Additionally, when the specific entry is used, the command line generated is not correct.

For instance, using the p27 entry the following command is run in the terminal:

```
tox -e py27 /home/bthornto/github/tox_bug
```

Can the test tree be generated by a `tox -l` command ?
Can the command line run simply be `tox -e <envname>`?

Being able to run any tox environment from the test tree would be incredibly powerful and useful.

Thanks!

Brad (cidrblock)

**Comment by cidrblock** [comment]
```
(venv) m910x ➜  tox_bug tox -l
py27
py36
```
sample output from `tox -l` for the above

**Comment by s-hemer** [comment]
With a little more "complex" envlist like
```ini
[tox]
envlist = {py38,py39,py310}-{lint,test,doc,requirements}
```
, the testing tree is completely empty (beside the expandable tox.ini of course), although `tox -l`/`tox -a` lists them correctly:
```bash
py38-lint
py38-test
py38-doc
py38-requirements
py39-lint
py39-test
py39-doc
py39-requirements
py310-lint
py310-test
py310-doc
py310-requirements
```
Command palette yet lists this correctly.

**Comment by ssbarnea** [comment]
@The-Compiler Any chance you could avoid parsing the tox.ini file yourself and use tox "api" (`tox -la`) to gather the list of available environments? The way expansion works makes it close to impossible to re-implement by anyone else than tox itself, as plugins can add their own contributions.

**Comment by The-Compiler** [comment]
That's what we already do in general:

https://github.com/The-Compiler/vscode-python-tox/blob/34b8f0d139110f46c00d553d61c3ce866b60120d/src/run.ts#L9-L12

(except with `tox -a`, because `-l` would only list the ones under `envlist`, but that's just the [default list of environments](https://tox.wiki/en/latest/config.html#envlist)).

Though for the annotations inside the file, the annotation *needs* to be with the `[sections]` in the ini file, so it does need to be parsed manually. Things like expansions indeed are not expanded there. It's unclear to me if we could generate additional test items without a location inside the file for other environments.

However, personally I don't use this part of the extension (it was contributed by @starlord-daniel), and my time to spend on this project is currently rather limited, so it'll take a while (if it's fixable at all).

---

As for the second part, that's a duplicate of #31.

**Comment by s-hemer** [comment]
> That's what we already do in general:
> 
> https://github.com/The-Compiler/vscode-python-tox/blob/34b8f0d139110f46c00d553d61c3ce866b60120d/src/run.ts#L9-L12
> 
> (except with `tox -a`, because `-l` would only list the ones under `envlist`...

I'd suggest to use the newer `tox list` command (with `--no-desc`) here. `-a` and `-l` are legacy. `tox list --no-desc -d` would be the replacement for `-l`

**Comment by s-hemer** [comment]
> Any chance you could avoid parsing the tox.ini file yourself

> That's what we already do in general

https://github.com/The-Compiler/vscode-python-tox/blob/34b8f0d139110f46c00d553d61c3ce866b60120d/src/testController.ts#L111-L120

Here, the own parsing happens and this does explicitly
https://github.com/The-Compiler/vscode-python-tox/blob/34b8f0d139110f46c00d553d61c3ce866b60120d/src/testController.ts#L131

Maybe, instead of a separate parsing it should instead try to locate the tests from `tox list` output? And if it can't locate them, at least add them to the test view.

**Comment by The-Compiler** [comment]
> I'd suggest to use the newer `tox list` command (with `--no-desc`) here. `-a` and `-l` are legacy. `tox list --no-desc -d` would be the replacement for `-l`

As long as it still works in tox 4, I'll continue to use `-a`. Everything else would just drop support for tox 3 with no benefits, while tox 3 still is rather widespread (currently still e.g. the default even [on Archlinux](https://archlinux.org/packages/?sort=&q=python-tox&maintainer=&flagged=), and certainly on things like Debian/Ubuntu too).

**Comment by s-hemer** [comment]
Don't know if it works on tox 4, I am on Ubuntu 22.04 with tox 3.21.4 and the new variants work.

**Comment by The-Compiler** [comment]
You're not running tox 3 then, because this [quite simply didn't exist](https://tox.wiki/en/3.27.1/config.html#cli) with tox 3:

```
$ tox --version
3.26.0 imported from /usr/lib/python3.10/site-packages/tox/__init__.py

$ tox list --no-desc
usage: tox [--version] [-h] [--help-ini] [-v] [-q] [--showconfig] [-l] [-a] [-c CONFIGFILE] [-e envlist] [--devenv ENVDIR] [--notest] [--sdistonly] [--skip-pkg-install] [-p [VAL]] [-o]
           [--parallel--safe-build] [--installpkg PATH] [--develop] [-i URL] [--pre] [-r] [--result-json PATH] [--discover PATH [PATH ...]] [--hashseed SEED] [--force-dep REQ]
           [--sitepackages] [--alwayscopy] [--no-provision [REQUIRES_JSON]] [-s [val]] [--workdir PATH]
           [args ...]
tox: error: unrecognized arguments: --no-desc

$ tox list          
GLOB sdist-make: ...
python inst-nodeps: ...
python installed: ...
python run-test-pre: ...
python run-test: commands[0] | pytest
```

**Comment by s-hemer** [comment]
As I  wrote, no venv active:
```
$ tox list --no-desc
py38-lint
py38-test
py38-doc
py38-requirements
py39-lint
py39-test
py39-doc
py39-requirements
py310-lint
py310-test
py310-doc
py310-requirements
$ tox --version
3.21.4 imported from /usr/lib/python3/dist-packages/tox/__init__.py
$ apt-cache madison tox
       tox | 3.21.4-1ubuntu1 | http://de.archive.ubuntu.com/ubuntu jammy/universe amd64 Packages
```
But indeed, while trying to solve the issue, I stumbled upon a strange behavior of the plugin within VSCode (need to investigate deeper): some calls seem to work fine and some calls seem to use another, old version of tox (error messages listing the help similar to yours).

**Comment by s-hemer** [comment]
Ok, forget about what I just wrote:
There exists a `.tox/.tox/lib/python3.10/site-packages/tox/` which looks like of version 4.4.11. And a short test in a naked folder shows that due to the `minversion = v4.0` in my tox.ini, tox just installs the newest version itself (in every folder, I tried to get over it but I hate this auto-dependency drawing/venv/dependency hell ... Debian package/system install or ~~die~~ DIY)

**Comment by s-hemer** [comment]
So back to constructive: 
no matter if `tox list` or ` tox -l/-a`, it may be a clean way to define these commands once for all calls (like in ToxTaskProvider:94). Similar, I'd suggest to re-use the `getToxEnvs` from run, then maybe it is sufficient to just call this wherever the list of environments is needed. Yet, I did not succeed due to this mixed use of tox versions.

**Comment by The-Compiler** [comment]
Agreed that `toxTaskProvider.ts` should probably reuse `getToxEnvs`. There [were a lot of contributions](https://github.com/The-Compiler/vscode-python-tox/pulls?q=is%3Apr+is%3Aclosed) to this extension in parallel, in a relatively short amount of time - while I tried my best to refactor and clean up things afterwards, I might have missed that one.

But again, "getting all test environments" and "getting the test environment corresponding to a given header line in a `tox.ini`" are two different beasts. From what I understand, the aim of the testing integration was the latter (see #11).

It's unclear to me whether VS Code allows creating [a test item](https://code.visualstudio.com/api/references/vscode-api#TestItem) without specifying a [range](https://code.visualstudio.com/api/references/vscode-api#Range) where in the file it is. If that's possible, then we could use `tox -a` *in addition* to the manual parsing to add "dummy" items for all tox environments we can't clearly associate with a place in the `tox.ini`. If that's not possible, I feel like this is a wontfix, and the task/command interface should just be used for cases like this instead.

**Comment by s-hemer** [comment]
I totally got that. Testing integration from #11 did exclude the handling of  "tox permutations for now" (see the comment in testController.ts:133) and this would be exactly the point to attack: retrieve lines of the environments definitions from `tox -a`. For the first shot, I would try to add them to line 0 if I can't find them directly with the regex as is (then they are at least listed in the tree overview hopefully). Second would be a correct resolution of the permutations of course. 
This way we can ensure that it doesn't find `testenv:`s that tox would refuse to run (which relates for me also to the security note in run about injecting malicious commands, doesn't it?).

**Comment by The-Compiler** [comment]
Ah, you mean starting from `tox -a` and then basically searching for `[envname]` in the file instead to find the line for it? That does seem like a great approach actually. Seems like an improvement over the status quo even without handling permutations, I suppose.

If adding them to line 0 if not found works, sure, that seems fine. I'm just not sure what VS Code does in situations like that.

As for the security note, I don't think it affects much. All of the justifications there still hold from what I can see. I'm assuming you mean the last one there, but the intention there was more to say "it's unlikely that this would break things (rather than be a security issue) in normal day to day usage".

**Comment by s-hemer** [comment]
I also do not see it very critical, yet it is one stage more, the tox-internal parsing, in between, and not just the `envName` found via the regex that may be run through the `testController`. Short: if tox can't parse it, it also won't run it and vice versa: if it can parse it, it may be safe to also run it.

**Comment by s-hemer** [comment]
I may need some time but I think I will come up with a suggestion soon.

**Comment by s-hemer** [comment]
Prototype looks promising, need some clean-up, then I will create a pull request.

**Comment by s-hemer** [comment]
> But indeed, while trying to solve the issue, I stumbled upon a strange behavior of the plugin within VSCode: some calls seem to work fine and some calls seem to use another version of tox.

When your system tox is v3, yet your tox.ini wants i.e. `minversion=v4.0` (and installs one, i.e. 4.4.x to `.tox`), it gets problematic as some parts of the plugin (can) just use the system version but others like the test calls in the project need to run/automatically run with the version present in .tox. Seems that I need to somehow add a check for the `minversion` in tox.ini
Especially, running environments from/on a specific folder, with v3 `tox -e <env> <path>`, does not work for v4 and needs to be `tox run -e <env> --root <path>`. `tox -e <env> --root <path>` works for v4 but not for v3, so at least for adding the path, a switch is needed.

**Comment by The-Compiler** [comment]
> > But indeed, while trying to solve the issue, I stumbled upon a strange behavior of the plugin within VSCode: some calls seem to work fine and some calls seem to use another version of tox.
> 
> When your system tox is v3, yet your tox.ini wants i.e. `minversion=v4.0` (and installs one, i.e. 4.4.x to `.tox`), it gets problematic as some parts of the plugin (can) just use the system version but others like the test calls in the project need to run/automatically run with the version present in .tox. Seems that I need to somehow add a check for the `minversion` in tox.ini

That seems like an odd approach to me - as mentioned, even the system-wide tox can be either v3 or v4.

I'd really encourage you to try and keep compatibilty with both versions rather than introducing this much additional complexity which is almost bound to break in some way.

> Especially, running environments from/on a specific folder, with v3 `tox -e <env> <path>`, does not work for v4 and needs to be `tox run -e <env> --root <path>`. `tox -e <env> --root <path>` works for v4 but not for v3, so at least for adding the path, a switch is needed.

Those do wildly different things. The former simply passes the path as `{posargs}` to the underlying command.

It's unclear to me what you're aiming at with the latter, or why it's needed - can you elaborate? There is `--workdir` that works with both, or you could use [the `cwd` option of `exec`](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) presumably. But I don't follow at all how any of this is connected to a fix for this issue.

**Comment by s-hemer** [comment]
> That seems like an odd approach to me - as mentioned, even the system-wide tox can be either v3 or v4.

An odd approach from tox to just silently install its own updated version? Agree. The minversion is not under my control.

> I'd really encourage you to try and keep compatibility with both versions

Definitively, yet this is what introduces this problem. I need to be able to get the environments from both tox versions (in parallel) as the plugin generates the (inital) list with the system version (that works with `tox -l` for both still), for the test tree it may already use the one from project's `.tox`, and the test runs then definitively use the one from `.tox` (if present).

> It's unclear to me what you're aiming at with the latter, or why it's needed - can you elaborate?

The code of the plugin right now contains some calls like v3 `tox -e <env> <path>`, yet due to tox potentially installing a v4 where this does not work anymore, it needs the compatibility with both versions, so either a call that works for both, or a switch.
`--workdir` is not working (in terms of giving the path as in the v3 call above), cmp tox v4 help:
```
 --workdir dir               tox working directory (if not specified will be the folder of the config file) (default:<path>/.tox -> from env var TOX_WORK_DIR)
```

Running the tox tests is not done with exec but with sending a command to terminal (either with `vscode.Terminal.sendText()` or `vscode.ShellExecution`.

I will try to clean up my stuff, then you may see from the code.

**Comment by The-Compiler** [comment]
> > That seems like an odd approach to me - as mentioned, even the system-wide tox can be either v3 or v4.
> 
> An odd approach from tox to just silently install its own updated version? Agree. The minversion is not under my control.

No, anything involving parsing `minversion` from the extension. A discussion about whether that behavior from tox makes sense or not seems vastly off-topic here.

> > I'd really encourage you to try and keep compatibility with both versions
> 
> Definitively, yet this is what introduces this problem. I need to be able to get the environments from both tox versions (in parallel) as the plugin generates the (inital) list with the system version (that works with `tox -l` for both still), for the test tree it may already use the one from project's `.tox`, and the test runs then definitively use the one from `.tox` (if present).

In what scenario would the two differ, and in which scenario would the system-wide one get the environment list? I tried both `tox -a` and `tox -l` with a system-wide tox 3 and `minversion = 4`, and they did provision tox 4 in `.tox` before printing the list.

> > It's unclear to me what you're aiming at with the latter, or why it's needed - can you elaborate?
> 
> The code of the plugin right now contains some calls like v3 `tox -e <env> <path>`, yet due to tox potentially installing a v4 where this does not work anymore, it needs the compatibility with both versions, so either a call that works for both, or a switch.

It never worked. It's a bug in the extension caused by a bad merge, see #31. It does not do what you think it does with tox 3.

> `--workdir` is not working (in terms of giving the path as in the v3 call above), cmp tox v4 help:
> 
> ```
>  --workdir dir               tox working directory (if not specified will be the folder of the config file) (default:<path>/.tox -> from env var TOX_WORK_DIR)
> ```

My bad. My tox 4.4.12 only says "tox working directory (if not specified will be the folder of the config file) (default: None)" which then seems both confusing and wrong.

There is also `-c` which we might be able to use, though not sure what that will do to relative paths in the `tox.ini`.

> Running the tox tests is not done with exec but with sending a command to terminal (either with `vscode.Terminal.sendText()` or `vscode.ShellExecution`.

Same idea then, the `vscode.createTerminal` function [has a `cwd` option](https://code.visualstudio.com/api/references/vscode-api#TerminalOptions) which we already use:

https://github.com/The-Compiler/vscode-python-tox/blob/34b8f0d139110f46c00d553d61c3ce866b60120d/src/utils.ts#L35

Though I don't know your code in detail, so...

> I will try to clean up my stuff, then you may see from the code.

...I'[l indeed wait for that until I say more. For what it's worth, I'd really appreciate having a simple but working version in first (less code -> much easier to review), and then we can take it from there. It seems like it'd be an improvement over the status quo either way.

**Comment by s-hemer** [comment]
Well, here it is. Last commit is just the state that "works for me" and potentially work with bot v3 and v4 tox. As basis for discussions.

> No, anything involving parsing minversion

I'd also like to circumvent that. Discussing any "oddness" is just relevant as long we have to deal with that. Truly odd is that tox does not give a hint when calling --version (which made me believe that I am on v3 all the time, resulting in this elongated discussion now hidden :smile: ).

> In what scenario would the two differ, and in which scenario would the system-wide one get the environment list?

I do not remember how/when I had that impression, yet this made me dig deeper. Otherwise just using v4 commands would have worked. But maybe it just was a strange constellation. Try to cover that in a test setup.

>It never worked.

Ok. Then it truly misled me. If default is none for `--workdir` (which I used now but giving the path to `.tox`); I need to investigate where `TOX_WORK_DIR` was set in my setup.

**Comment by s-hemer** [comment]
Ah, well, first additonal simple artificial test setup showed that this is not the final solution :disappointed:
